<a href="https://colab.research.google.com/github/Fushiguro19/Python-basics-practice/blob/Python-colab/Python%20learning%20colab/Python_assignment1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

1) Write a function that inputs a number and prints the multiplication table of that number

In [ ]:
def table(x,y) :
    i = 0
    while i<=y:
        a = x*i
        print(f"\n{x} * {i} = {a}.")
        i = i+1
    return("\nEND")
    

a = int(input("enter a number : "))
b = int(input("\nenter the number upto which multiplication table is needed : "))


print (F"\nMultiplication Table of {a} : .") 

print(table(a,b))

enter a number : 5

enter the number upto which multiplication table is needed : 10

Multiplication Table of 5 : .

5 * 0 = 0.

5 * 1 = 5.

5 * 2 = 10.

5 * 3 = 15.

5 * 4 = 20.

5 * 5 = 25.

5 * 6 = 30.

5 * 7 = 35.

5 * 8 = 40.

5 * 9 = 45.

5 * 10 = 50.

END


2) Write a program to print twin primes less than 1000. If two consecutive odd numbers are
both prime then they are known as twin primes

In [ ]:
#list of prime numbers upto 1000

prime = []

for n in range (1,1000) :
    for i in range (2, ((n//2)+1)) :
        if (n%i == 0) :
            break
        
    else : 
        if n!= 1 :
            prime.append(n)


#picking out twin primes from the list

twins = [(prime[i],prime[i+1]) for i in range (0,(len(prime)-1)) if (prime[i+1]-prime[i] == 2)]
    
print (twins)


[(3, 5), (5, 7), (11, 13), (17, 19), (29, 31), (41, 43), (59, 61), (71, 73), (101, 103), (107, 109), (137, 139), (149, 151), (179, 181), (191, 193), (197, 199), (227, 229), (239, 241), (269, 271), (281, 283), (311, 313), (347, 349), (419, 421), (431, 433), (461, 463), (521, 523), (569, 571), (599, 601), (617, 619), (641, 643), (659, 661), (809, 811), (821, 823), (827, 829), (857, 859), (881, 883)]


3) Write a program to find out the prime factors of a number. Example: prime factors of 56 -
2, 2, 2, 7

In [ ]:
n = int(input("enter the number ="))

#empty list for all the prime factors
pf = []

pf.append(1)
#ruling out all possibilities of prime factor 2
while (n%2 == 0) :
    pf.append(2)
    n = n//2

    
#all prime factors less than sqrt of n
for i in range (3, int(n**(1/2))+1, 2) :
    while (n%i == 0) :
        pf.append(i)
        n = n//i

#to include the possibility of a prime factor greater than sqrt of n coz there is only 1(if any)
if n > 2 :
    pf.append(n)
    

print (pf)
    

enter the number =1035
[1, 3, 3, 5, 23]


4) Write a program to implement these formulae of permutations and combinations.
Number of permutations of n objects taken r at a time: p(n, r) = n! / (n-r)!. Number of
combinations of n objects taken r at a time is: c(n, r) = n! / (r!*(n-r)!) = p(n,r) / r!

In [ ]:
n = int(input("enter the number of objects,n ="))
r = int(input("no of objects taken at a time,r ="))
p = 1

for i in range ((n-r+1),(n+1)) :
    p = p*i
print(f"Permutations of {n} objects taken {r} at a time P({n},{r}) = {p} ")

q = 1
for i in range (1,(r+1)) :
    q = q*i

c = p//q
print(f"Combinations of {n} objects taken {r} at a time C({n},{r}) = {c} ")

enter the number of objects,n =5
no of objects taken at a time,r =3
Permutations of 5 objects taken 3 at a time P(5,3) = 60 
Combinations of 5 objects taken 3 at a time C(5,3) = 10 


5) Write a function that converts a decimal number to binary number

In [ ]:
n = int(input("enter the number,n ="))
num = n
d = []

while (n>0) :
    p = (n%2)
    n = n//2
    d.append(p)

d.reverse()
    
print (f"the binary equivalent of {num} = ")

for j in d :
    print (j, end = " ")

enter the number,n =17
the binary equivalent of 17 = 
1 0 0 0 1 

In [ ]:
def binary(n) :
    if n >0 :
        binary(n//2)
        print(n%2, end = "")
         
       
x = int(input("enter the number = "))
binary(x)

##learnt from the below link
##https://www.programiz.com/python-programming/examples/decimal-binary-recursion

enter the number = 10
1010

6) Write a function cubesum() that accepts an integer and returns the sum of the cubes of
individual digits of that number. Use this function to make functions PrintArmstrong() and
isArmstrong() to print Armstrong numbers and to find whether is an Armstrong number.

In [ ]:

def cubesum(n) :
    ##https://www.programiz.com/python-programming/examples/armstrong-number
    s = 0
    while n > 0 :
        s = s + ((n%10)**3)
        n = n//10
    return(s)

def isArmstrong(x) :

    if x == cubesum(x):
        return("This is an Armstrong Number")
    else :
        return("Not an Armstrong number")

def PrintArmstrong(a):
    q = []
    for i in range(a):
        if i == cubesum(i):

            q.append(i)
    return(q)
    
a = int(input("enter a number : "))
print("The sum of the cube of its digits is : ",cubesum(a))
print(isArmstrong(a))
print("Armstrong numbers upto ", a, "are : ", PrintArmstrong(a))

## Having a DOUBT in this problem

enter a number : 220
The sum of the cube of its digits is :  16
Not an Armstrong number
Armstrong numbers upto  220 are :  [0, 1, 153]


7) Write a function prodDigits() that inputs a number and returns the product of digits of that
number.

In [ ]:
####https://stackoverflow.com/questions/21270320/turn-a-single-number-into-single-digits-python

def prodDigits(x) :
    
    p = [int(a) for a in str(x)]
    m = 1
    for i in p :
        m = m*i
    return(m)

n = int(input("enter a number"))

print("Product of the digits of ", n, " is : ", prodDigits(n))

enter a number89
Product of the digits of  89  is :  72


8) If all digits of a number n are multiplied by each other repeating with the product, the one
digit number obtained at last is called the multiplicative digital root of n. The number of
times digits need to be multiplied to reach one digit is called the multiplicative
persistance of n.
Example: 86 -> 48 -> 32 -> 6 (MDR 6, MPersistence 3)
341 -> 12->2 (MDR 2, MPersistence 2)
Using the function prodDigits() of previous exercise write functions MDR() and
MPersistence() that input a number and return its multiplicative digital root and
multiplicative persistence respectively

In [ ]:
def prodDigits(x) :
    
    p = [int(a) for a in str(x)]
    m = 1
    for i in p :
        m = m*i
    return(m)

def MDR(n) :
    
    def Mper(n):
        if len(str(n)) == 1 :
            print(n)
        else :
            s = 0
            while len(str(n))>1 :
                n = prodDigits(n)
                s += 1
            print(n, s)
    Mper(n)
    
    
        
t = int(input("enter a number : "))     
print("The MDR, Mpersistence are : ")
MDR(t)

##https://www.geeksforgeeks.org/python-inner-functions/

enter a number : 189
The MDR, Mpersistence are : 
4 3


9) Write a function sumPdivisors() that finds the sum of proper divisors of a number. Proper
divisors of a number are those numbers by which the number is divisible, except the
number itself. For example proper divisors of 36 are 1, 2, 3, 4, 6, 9, 18

In [ ]:
def sumPdivisors(n) :
    s = 0
    for i in range(1,n):
        if (n%i) == 0 :
            s += i
    return(s)

t = int(input("enter a number : "))
print("Sum of proper divisors is = ",sumPdivisors(t))
            

enter a number : 22
Sum of proper divisors is =  14


10) A number is called perfect if the sum of proper divisors of that number is equal to the
number. For example 28 is perfect number, since 1+2+4+7+14=28. Write a program to
print all the perfect numbers in a given range

In [ ]:
def sumPdivisors(n) :
    s = 0
    for i in range(1,n):
        if (n%i) == 0 :
            s += i
    return(s)

def perfect(p) :
    q = []
    for i in range(0,p) :
        if i == sumPdivisors(i) :
            q.append(i)
    return(q)
        

t = int(input("enter a number : "))
print("Perfect numbers from 0 to ", t, " are : ")
perfect(t)
            

enter a number : 220
Perfect numbers from 0 to  220  are : 


[0, 6, 28]

11) Two different numbers are called amicable numbers if the sum of the proper divisors of
each is equal to the other number. For example 220 and 284 are amicable numbers.
Sum of proper divisors of 220 = 1+2+4+5+10+11+20+22+44+55+110 = 284
Sum of proper divisors of 284 = 1+2+4+71+142 = 220
Write a function to print pairs of amicable numbers in a range


In [ ]:
def sumPdivisors(n) :
    s = 0
    for i in range(1,n):
        if (n%i) == 0 :
            s += i
    return(s)


def ami(n):
  for i in range(0,(n+1)) :
    for j in range((i+1), (n+1)):
      if ((j == sumPdivisors(i)) and (i == sumPdivisors(j))) :
        print(i,j)
    

t = int(input("enter a number"))
ami(t)
        
##https://www.geeksforgeeks.org/pairs-amicable-numbers/

enter a number1000
220 284


12) Write a program which can filter odd numbers in a list by using filter function

In [ ]:
x = int(input("enter a number : "))
q = [i for i in range(0,(x+1))]
odd_nos = filter(lambda i : (i%2 != 0),q )
print("Odd numbers till ", x , " are : ", list(odd_nos))


##https://www.pythontutorial.net/python-basics/python-filter-list/

enter a number : 20
Odd numbers till  20  are :  [1, 3, 5, 7, 9, 11, 13, 15, 17, 19]


13) Write a program which can map() to make a list whose elements are cube of elements in
a given list

In [ ]:
x = int(input("enter a number : "))
q = [i for i in range(0,(x+1))]
cubes = map(lambda i : (i**3),q )
print("Cubes of numbers till ", x , " are : ", list(cubes))

##https://www.geeksforgeeks.org/python-map-function/

enter a number : 10
Cubes of numbers till  10  are :  [0, 1, 8, 27, 64, 125, 216, 343, 512, 729, 1000]


14) Write a program which can map() and filter() to make a list whose elements are cube of
even number in a given list

In [ ]:
x = int(input("enter a number : "))
q = [i for i in range(0,(x+1))]
FM = map(lambda n : n**3, (filter(lambda n : (n%2 == 0),q)))
print(list(FM))

enter a number : 20
[0, 8, 64, 216, 512, 1000, 1728, 2744, 4096, 5832, 8000]
